### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by enjmiah**
Mostly formatting and documentation improvements.

Also fix an error when there is a zero along the diagonal for `forsub` and `backsub`.  Now it just answers `false` in that case.

Also add some more tests

### CODE REVIEW

File: matrixlog.pl [hunk 1]
```diff
@@ -193,87 +239,114 @@
...
+  functor(AElements,elements,Size),
+  functor(LElements,elements,Size),
+  functor(UElements,elements,Size),
+  lu_row(0, N, AElements, LElements, UElements),
+  L = matrix(N, N, LElements),
+  U = matrix(N, N, UElements).
+
+% Compute a single row of Lower and Upper.
+lu_row(N, N, _As, _Ls,   _Us) :- !.
+lu_row(J, N, As,  Lower, Upper) :-
```
Comment (JGAJ) [inline matrixlog.pl]: I might call this lu_col sense j is the column index

File: matrixlog.pl [hunk 2]
```diff
@@ -193,87 +239,114 @@
...
+  Jplus #= J + 1,
+  IndexJJ #= Jplus + J * N,
+  setarg(IndexJJ, Lower, 1), % set L[Jplus,Jplus] = 1
+  lu_element(0,Jplus,N,As,Lower,Upper),
+  lu_row(Jplus,N,As,Lower,Upper).
+
+% Compute a single element of Lower and Upper.
+lu_element(N, _J, N, _As, _Ls,   _Us) :- !.
```
Comment (JGAJ) [inline matrixlog.pl]: this I would call lu_row because it iterates over the rows

File: matrixlog.pl [hunk 3]
```diff
@@ -193,87 +239,114 @@
...
-      ValLower is (Av-Sum)/Ujj, 
+      ValLower is (Av-Sum)/Ujj,
       setarg(Index,Lower,ValLower),
+      % and zero out the corresponding value of L
       setarg(Index,Upper,0))),
-  lu_i(Iplus,J,N,As,Lower,Upper).
+  lu_element(Iplus,J,N,As,Lower,Upper).
+
+% Computes sum of Lower[I, K] * Upper[K, J] for K = 1..Stop-1
+sum_l_ik_u_kj(Stop,I,J,N,Lower,Upper,Sum) :-
```
Comment (JGAJ) [inline matrixlog.pl]: lu_elem_helper?
add comment %U[I][J] is A[I][J] - Sum, L[I][J] = (A[I][J]-Sum)/U[J][J], for I,J in Upper and lower triangles halves of A respectively, unless I=J in which case L[I][I] =1

File: matrixlog.pl [hunk 4]
```diff
@@ -406,66 +479,130 @@
...
-              X = matrix(2,1,elements(1,2)),
-              assertion(solve(A,X,B)).
-
-error(zero_divisor):-lu(matrix(2,2,elements(0,1,1,0)),L,U),
-              valid_matrix(L),valid_matrix(U). 
-error(zero_divisor):- solve(matrix(2,2,elements(0,1,1,0)),
-                      matrix(2,1,elements(2,1)),
-                      matrix(2,1,elements(1,2))).
-:- end_tests(matrixlog).
+test(forsub) :-
```
Comment (JGAJ) [inline matrixlog.pl]: Thanks for writing these tests, much better then what I had.

**Review by JGAJ — COMMENTED** [review]
Looks good, I appreciate the extra tests, mostly just comments about naming.

**Comment by enjmiah** [comment]
Made the changes you requested. Merging now because it's getting late...

Let me know if you want to make further changes.

**Pull request merged**